File: backend/src/main/java/by/minilooth/diploma/models/bean/spareparts/SparePart.java
package by.minilooth.diploma.models.bean.spareparts;

import by.minilooth.diploma.models.api.AbstractEntity;
import by.minilooth.diploma.models.bean.catalog.Category;
import by.minilooth.diploma.models.bean.catalog.Group;
import by.minilooth.diploma.models.bean.catalog.Subcategory;
import by.minilooth.diploma.models.bean.common.Image;
import by.minilooth.diploma.models.bean.cart.CartItem;
import by.minilooth.diploma.models.bean.deals.Trade;
import by.minilooth.diploma.models.bean.stores.Availability;
import by.minilooth.diploma.models.bean.vehicle.Generation;
import by.minilooth.diploma.models.bean.vehicle.Make;
import by.minilooth.diploma.models.bean.vehicle.Model;
import lombok.*;

import javax.persistence.*;
import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "spare_part")
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Data
@EqualsAndHashCode(callSuper = false)
public class SparePart extends AbstractEntity {

    @Column(name = "name", nullable = false)
    private String name;

    @ManyToOne
    @JoinColumn(name = "manufacturer_id", nullable = false)
    private Manufacturer manufacturer;

    @Column(name = "article", nullable = false)
    private String article;

    @Column(name = "description", columnDefinition = "LONGTEXT")
    private String description;

    @Column(name = "purchase_price", nullable = false)
    private Float purchasePrice;

    @Column(name = "retail_price", nullable = false)
    private Float retailPrice;

    @OneToMany(mappedBy = "sparePart", targetEntity = Characteristic.class, cascade = {CascadeType.PERSIST, CascadeType.MERGE}, orphanRemoval = true)
    @EqualsAndHashCode.Exclude
    @Builder.Default
    private Set<Characteristic> characteristics = new LinkedHashSet<>();

    @OneToMany(mappedBy = "sparePart", targetEntity = Availability.class, cascade = {CascadeType.PERSIST, CascadeType.MERGE}, orphanRemoval = true)
    @EqualsAndHashCode.Exclude
    @Builder.Default
    private Set<Availability> availabilities = new LinkedHashSet<>();

    @OneToMany(mappedBy = "sparePart", targetEntity = CartItem.class, cascade = {CascadeType.PERSIST, CascadeType.MERGE}, orphanRemoval = true)
    @EqualsAndHashCode.Exclude
    @Builder.Default
    private Set<CartItem> cartItems = new LinkedHashSet<>();

    @OneToMany(mappedBy = "sparePart", targetEntity = Trade.class, cascade = {CascadeType.PERSIST, CascadeType.MERGE}, orphanRemoval = true)
    @EqualsAndHashCode.Exclude
    @Builder.Default
    private Set<Trade> trades = new LinkedHashSet<>();

    @ManyToOne
    @JoinColumn(name = "make_id", nullable = false)
    private Make make;

    @ManyToOne
    @JoinColumn(name = "model_id", nullable = false)
    private Model model;

    @ManyToOne
    @JoinColumn(name = "generation_id")
    private Generation generation;

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne
    @JoinColumn(name = "subcategory_id", nullable = false)
    private Subcategory subcategory;

    @ManyToOne
    @JoinColumn(name = "group_id")
    private Group group;

    @OneToOne(cascade = CascadeType.MERGE, fetch = FetchType.LAZY, orphanRemoval = true)
    @JoinTable(name = "spare_part_image",
            joinColumns =
                    { @JoinColumn(name = "spare_part_id", referencedColumnName = "id") },
            inverseJoinColumns =
                    { @JoinColumn(name = "image_id", referencedColumnName = "id") })
    private Image image;

    public void setCharacteristics(Set<Characteristic> characteristics) {
        this.characteristics.clear();
        if (Objects.nonNull(characteristics)) {
            this.characteristics.addAll(characteristics);
        }
    }

    public void setAvailabilities(Set<Availability> availabilities) {
        this.availabilities.clear();
        if (Objects.nonNull(availabilities)) {
            this.availabilities.addAll(availabilities);
        }
    }

    public void setCartItems(Set<CartItem> cartItems) {
        this.cartItems.clear();
        if (Objects.nonNull(cartItems)) {
            this.cartItems.addAll(cartItems);
        }
    }

    public void setTrades(Set<Trade> trades) {
        this.trades.clear();
        if (Objects.nonNull(trades)) {
            this.trades.addAll(trades);
        }
    }

}

File: frontend/types/spareparts/sparePart.ts
import {Modification} from "types/spareparts/modification";
import {Store} from "types/stores/store";
import {Manufacturer} from "types/spareparts/manufacturer";
import {Make} from "types/spareparts/vehicle/make";
import {Model} from "types/spareparts/vehicle/model";
import {Generation} from "types/spareparts/vehicle/generation";
import {Category} from "types/spareparts/catalog/category";
import {Subcategory} from "types/spareparts/catalog/subcategory";
import {Group} from "types/spareparts/catalog/group";
import {SortDirection} from "types/common/sort-direction";
import {Image} from "types/image";

export interface SparePart {
  id: number;
  name: string;
  manufacturer: Manufacturer;
  article: string;
  description: string;
  purchasePrice: number;
  purchasePriceUsd: number;
  purchasePriceEur: number;
  retailPrice: number;
  retailPriceUsd: number;
  retailPriceEur: number;
  characteristics: Characteristic[],
  availabilities: Availability[],
  totalQuantity: number;
  make: Make,
  model: Model,
  generation: Generation;
  category: Category;
  subcategory: Subcategory;
  group: Group;
  image: Image;
  createdAt: Date;
  updatedAt: Date;
}

export interface Characteristic {
  id: CharacteristicKey;
  modification: Modification;
  value: string;
}

export interface CharacteristicKey {
  sparePartId: number;
  modificationId: number;
}

export interface Availability {
  id: AvailabilityKey;
  store: Store;
  quantity: number;
}

interface AvailabilityKey {
  storeId: number;
  sparePartId: number;
}

export interface SparePartList {
  spareParts: SparePart[];
  pages: number;
}

export enum SparePartSort {
  NAME,
  MANUFACTURER,
  ARTICLE,
  PURCHASE_PRICE,
  RETAIL_PRICE,
  MAKE,
  MODEL,
  GENERATION,
  CATEGORY,
  SUBCATEGORY,
  GROUP
}


export interface ProcessSparePart {
  name: string;
  manufacturer: Manufacturer;
  article: string;
  purchasePrice: number;
  retailPrice: number;
  characteristics: Characteristic[],
  make: Make,
  model: Model,
  generation: Generation;
  category: Category;
  subcategory: Subcategory;
  group: Group;
  image: Image;
}

export interface SparePartSortItem {
  field: SparePartSort;
  label: string;
  query: string;
}

export const SparePartSortItems: SparePartSortItem[] = [
  {field: SparePartSort.NAME, label: 'Название', query: 'name'},
  {field: SparePartSort.MANUFACTURER, label: 'Производитель', query: 'manufacturerId'},
  {field: SparePartSort.ARTICLE, label: 'Артикул', query: 'article'},
  {field: SparePartSort.PURCHASE_PRICE, label: 'Закупочная цена', query: 'purchasePrice'},
  {field: SparePartSort.RETAIL_PRICE, label: 'Розничная цена', query: 'retailPrice'},
  {field: SparePartSort.MAKE, label: 'Марка', query: 'makeId'},
  {field: SparePartSort.MODEL, label: 'Модель', query: 'modelId'},
  {field: SparePartSort.GENERATION, label: 'Поколение', query: 'generationId'},
  {field: SparePartSort.CATEGORY, label: 'Категория', query: 'categoryId'},
  {field: SparePartSort.SUBCATEGORY, label: 'Подкатегория', query: 'subcategoryId'},
  {field: SparePartSort.GROUP, label: 'Группа', query: 'groupId'}
]

export interface SparePartFilter {
  sort: string;
  sortDirection: string;
  manufacturerId: number;
  article: string;
  purchasePriceFrom: number;
  purchasePriceTo: number;
  retailPriceFrom: number;
  retailPriceTo: number;
  availabilityFrom: number;
  availabilityTo: number;
  makeId: number;
  modelId: number;
  generationId: number;
  categoryId: number;
  subcategoryId: number;
  groupId: number;
  search: string;
  page: number;
}
